### PR TITLE
mCODE Base Functionality - From Scratch

### PR DESCRIPTION
**Pull request by RobertScalfani**
This Pull Request adds the base functionality of mCODE, starting the inferno-mcode project from scratch. It adds:
- Read and Validate tests for every mCODE profile
- Auto-generation of mCODE Read and Validate tests from Capability Statements

---

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,35 +1,54 @@
...
 
-Inferno is an open source tool that tests whether patients can access their health data through a standard interface.
-It makes HTTP(S) requests to test your server's conformance to authentication, authorization, and FHIR content standards and reports the results back to you.
+# Inferno for mCODE
 
-This application creates test clients that exercise the range of requirements of a [SMART on FHIR](http://smarthealthit.org/smart-on-fhir/) server. These clients have tests for the [SMART App Launch Framework](http://www.hl7.org/fhir/smart-app-launch/), [Argonaut](http://hl7.org/fhir/DSTU2/argonaut/argonaut.html) Implementation Guides, and [FHIR DSTU2](http://hl7.org/fhir/DSTU2/index.html).
+## The Situational Awareness for Novel Epidemic Response Implementation Guide Testing
 
-## Using Inferno
+**Inferno mCODE Edition** performs API conformance testing for the [Sminimal Common Oncology Data Elements (mCODE) Release 1 - US Realm | STU1 Implementation Guide](http://hl7.org/fhir/us/mcode/index.html).
```
Comment (radamson) [inline README.md]:
```suggestion
**Inferno mCODE Edition** performs API conformance testing for the [minimal Common Oncology Data Elements (mCODE) Release 1 - US Realm | STU1 Implementation Guide](http://hl7.org/fhir/us/mcode/index.html).
```

File: README.md [hunk 2]
```diff
@@ -1,35 +1,54 @@
...
-[![Build Status](https://travis-ci.org/onc-healthit/inferno.svg?branch=master)](https://travis-ci.org/onc-healthit/inferno)
+[![Build
+Status](https://travis-ci.org/onc-healthit/inferno.svg?branch=master)](https://travis-ci.org/onc-healthit/inferno-program)
 
-Inferno is an open source tool that tests whether patients can access their health data through a standard interface.
-It makes HTTP(S) requests to test your server's conformance to authentication, authorization, and FHIR content standards and reports the results back to you.
+# Inferno for mCODE
 
-This application creates test clients that exercise the range of requirements of a [SMART on FHIR](http://smarthealthit.org/smart-on-fhir/) server. These clients have tests for the [SMART App Launch Framework](http://www.hl7.org/fhir/smart-app-launch/), [Argonaut](http://hl7.org/fhir/DSTU2/argonaut/argonaut.html) Implementation Guides, and [FHIR DSTU2](http://hl7.org/fhir/DSTU2/index.html).
+## The Situational Awareness for Novel Epidemic Response Implementation Guide Testing
```
Comment (radamson) [inline README.md]: anything SANER related should be removed.  Please take another sweep through the repo.

File: config.yml [hunk 3]
```diff
@@ -35,65 +35,90 @@ autorun_capability: false
...
 include_extras: true
 
-badge_text: Community
+badge_text: mCODE EDITION
 
 # Resource validator options: must be one of "internal" or "external". external_resource_validator_url is only used if resource_validator is set to external.
-resource_validator: internal
-external_resource_validator_url: http://validator_service:4567
+resource_validator: external
+external_resource_validator_url: http://localhost:8081
```
Comment (radamson) [inline config.yml]: this should probably stay as `http://validator_service:45467` so that the docker version works out of the box.  Make sure to test that `docker-compose up` still works.  You may have to add the volume mount to the docker-compose file.

File: generator/mcode/create_test.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        create_test = {
+          tests_that: "Server creates #{sequence[:resource]} resource with the #{sequence[:resource]} create interaction",
+          key: test_key,
+          index: sequence[:tests].length + 1,
+          link: 'http://hl7.org/fhir/us/mcode/index.html',
+          description: "A server #{interaction[:expectation]} support the #{sequence[:resource]} create interaction.",
+          optional: interaction[:expectation] != 'SHALL'
+        }
+        create_test[:test_code] = %(
+            #{sequence[:resource].downcase}_example = File.read(File.expand_path('./resources/saner/saner-#{sequence[:resource].downcase}-example.json'))
```
Comment (radamson) [inline generator/mcode/create_test.rb]: saner references should be corrected/removed.

File: generator/mcode/update_test.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        create_test = {
+          tests_that: "Server updates #{sequence[:resource]} resource with the #{sequence[:resource]} update interaction",
+          key: test_key,
+          index: sequence[:tests].length + 1,
+          link: 'http://hl7.org/fhir/us/mcode/index.html',
+          description: "A server #{interaction[:expectation]} support the #{sequence[:resource]} update interaction.",
+          optional: interaction[:expectation] != 'SHALL'
+        }
+        create_test[:test_code] = %(
+            #{sequence[:resource].downcase}_example = File.read(File.expand_path('./resources/saner/saner-#{sequence[:resource].downcase}-example.json'))
```
Comment (radamson) [inline generator/mcode/update_test.rb]: saner references should be corrected/removed

File: lib/app.rb [hunk 6]
```diff
@@ -24,6 +25,7 @@
 require_relative 'version'
 require_relative 'app/models'
 require_relative 'app/utils/terminology'
+require '/Users/rscalfani/Documents/code/inferno/inferno-saner/generator/mcode/mcode_generator'
```
Comment (radamson) [inline lib/app.rb]: this won't work on anyone else's machine.  Use a `require_relative` here to avoid reference to you as a user.

Comment (radamson) [reply lib/app.rb]: Also I don't think this `require` is necessary at all now that I think about it

**Comment by radamson** [comment]
Document how to generate the tests somewhere in the README.  Also it looks like there are some rubocop issues that Travis picked up which need to be addressed.

**Comment by radamson** [comment]
Add the instructions for how to run the generator to the README.

**Comment by RobertScalfani** [comment]
> Add the instructions for how to run the generator to the README.
